src/main/java/ch/njol/skript/lang/SkriptParser.java

```diff
@@ -237,19 +237,13 @@ public boolean hasTag(String tag) {
 							}
 							T element = info.getElementClass().newInstance();
 
-							if (element instanceof EventRestrictedSyntax eventRestrictedSyntax) {
-								Class<? extends Event>[] supportedEvents = eventRestrictedSyntax.supportedEvents();
-								if (!getParser().isCurrentEvent(supportedEvents)) {
-									Skript.error("'" + parseResult.expr + "' can only be used in " + supportedEventsNames(supportedEvents));
-									continue;
-								}
-							}
-							if (element instanceof ExperimentalSyntax experimentalSyntax) {
-								if (!experimentalSyntax.isSatisfiedBy(getParser().getExperimentSet()))
-									continue;
-							}
+							if (!checkRestrictedEvents(element, parseResult))
+								continue;
 
-							boolean success = element.init(parseResult.exprs, patternIndex, getParser().getHasDelayBefore(), parseResult);
+							if (!checkExperimentalSyntax(element))
+								continue;
+
+							boolean success = element.preInit() && element.init(parseResult.exprs, patternIndex, getParser().getHasDelayBefore(), parseResult);
```

Comment:
We have the opportunity to pass along information (such as the matched pattern). Do you think we should, or is it unlikely to be useful?

Reply:
I suppose it couldn't hurt, though I wanted to ensure the role of init stayed as it is rather than people moving stuff to pre-init unnecessarily.